update stored formats

cseed opened the pull request:
This PR improves and consolidates several that were pending.  It does a few things:

 - all values in the file formats are stored natively (encoded), not using json,
 - in particular, the globals and cols are not stored in the json, and not loaded when the read is performed by the user (but when it is executed by the backend).
 - MT parts (globals, rows, cols, entries) are all stored in separate directories and each is itself a valid Table file
 - extensions for MT and Table are no longer enforced
 - the metadata now stores an "RVDSpec" which is a recipe for reading an RVD.  This gives us the flexibility to define new RVD types (e.g. HashedRVD) or modify the existing classes (by forking and adding additional parameters, etc.)
 - the rows and entries are stored separately and zipped together on read.  If dropSamples = true, don't even load the entries.
 - created UnpartitionedRVD, cleaned up the RVD interface

 fixes: #2810

danking commented on src/main/scala/is/hail/rvd/RVD.scala:
```diff
@@ -1,27 +1,143 @@
+    val args = jargs.extract[JSONArgs]
+    UnpartitionedRVDSpec(Parser.parseStructType(args.row_type), args.part_files)
+  }
+}
```
You can use a custom serializer in json4s to avoid repeating the matrix parser
and TStruct parser stuff.

```scala
class TStructSerializer extends CustomSerializer[TStruct](format => (
  { case JString(s) => Parser.parseStructType(args.row_type) },
  { case x: TStruct => x.toString() }
))
```
you have to have the right formats in your implicit scope:
```scala
implicit val formats = DefaultFormats + new TStructSerializer
```

but then this is just:
```scala
object UnpartitionedRVDSpec {
  case class JSONArgs(row_type: String, part_files: Array[String])

  def extract(jargs: JValue): UnpartitionedRVDSpec = {
    val args = jargs.extract[JSONArgs]
    UnpartitionedRVDSpec(args.row_type, args.part_files)
  }
}

case class UnpartitionedRVDSpec(
  rowType: TStruct,
  partFiles: Array[String]) extends RVDSpec {
  def read(hc: HailContext, path: String): UnpartitionedRVD =
    new UnpartitionedRVD(rowType, hc.readRows(path, rowType, partFiles))

  def readLocal(hc: HailContext, path: String): IndexedSeq[Row] =
    RVDSpec.readLocal(hc, path, rowType, partFiles)

  def toJSON: JValue = JObject("name" -> JString("UnpartitionedRVDSpec"),
    "args" -> render(UnpartitionedRVDSpec.JSONArgs(rowType, partFiles)))
}
```

*but* we can get even better using TypeHints:

```scala
class TStructSerializer extends CustomSerializer[TStruct](format => (
  { case JString(s) => Parser.parseStructType(args.row_type) },
  { case x: TStruct => x.toString() }
))

implicit val formats = new DefaultFormats() {
  override val typeHints = ShortTypeHints(List(classOf[Foo]))
  override val typeHintFieldName = "name" } + new TStructSerializer

object UnpartitionedRVDSpec {
  def extract(jargs: JValue): UnpartitionedRVDSpec =
    jargs.extract[UnpartitionedRVDSpec]
}

case class UnpartitionedRVDSpec(
  row_type: TStruct,
  part_files: Array[String]) extends RVDSpec {
  def read(hc: HailContext, path: String): UnpartitionedRVD =
    new UnpartitionedRVD(rowType, hc.readRows(path, row_type, part_files))

  def readLocal(hc: HailContext, path: String): IndexedSeq[Row] =
    RVDSpec.readLocal(hc, path, row_type, part_files)

  def toJSON: JValue = render(this)
}
```

And the JSON should look like:

```javascript
{ "name" : "UnpartitionedRVDSpec"
, "row_type" : "... the toString'ed row type ..."
, "part_files" : ["part-00000", ...]
}
```

This seems to save us some of the boilerplate.

danking commented on src/main/scala/is/hail/table/Table.scala:
```diff
@@ -38,16 +39,53 @@ object SortColumn {
+    if (!FileFormat.version.supports(fileVersion))
+      fatal(s"incompatible file format when reading: $path\n  supported version: ${ FileFormat.version }, found $fileVersion")
+
+    val jmetadata = jv.extract[JSONTableMetadata]
```
same thing here, we could avoid the boilerplate and the JSONTableMetadata (assuming you're willing to live with underscores in scala or camelcase in JSON).

danking commented on src/main/scala/is/hail/utils/richUtils/RichRDD.scala:
```diff
@@ -195,6 +195,12 @@ class RichRDD[T](val r: RDD[T]) extends AnyVal {
+      val is = i.toString
+      assert(is.length <= d)
+      "part-" + StringUtils.leftPad(is, d, "0")
+    }
```
another occurrence

danking commented on src/main/scala/is/hail/methods/VEP.scala:
```diff
@@ -393,7 +393,7 @@ object VEP {
     val vepType: Type = if (csq) TArray(TString()) else vepSignature
 
-    val vepOrderedRVType = new OrderedRVType(
+    val vepOrderedRVType = new OrderedRVDType(
```
name needs to be updated to ...RVDType

catoverdrive commented on src/main/scala/is/hail/rvd/RVD.scala:
```diff
@@ -1,27 +1,143 @@
 trait RVD {
   self =>
-  def rowType: Type
+  // FIXME TStruct
```
what does this mean?

cseed commented on python/hail/api2/tests.py:
```diff
@@ -622,6 +622,21 @@ def test_sample_rows(self):
 
+    def test_read_stored_cols(self):
+        ds = self.get_vds()
+        ds = ds.annotate_globals(x = 'foo')
```
cols_table inherits the globals table from matrix table.  (Same with rows_table.)  I wanted to add non-trivial globals to make sure it got carried along correctly.

cseed commented:
Also, fixes: https://github.com/hail-is/hail/issues/2810

danking commented:
I believe fixes messages need to be in the PR message body to work, so I've added it there.

danking commented on src/main/scala/is/hail/io/RowStore.scala:
```diff
@@ -521,7 +523,95 @@ object RichRDDRegionValue {
+      val f = partFile(d, i)
+
+      val rowsPartPath = path + "/rows/rows/parts/" + f
+      val rowsOS = hConf.unsafeWriter(rowsPartPath)
```
this should be the serializable hadoop conf, right?

cseed replied:
It is?   Just above: `val hConf = sHConfBc.value.value`.

danking commented on src/main/scala/is/hail/io/RowStore.scala:
```diff
@@ -521,7 +523,95 @@ object RichRDDRegionValue {
+
+      entriesEN.writeByte(0)
+      entriesEN.flush()
+      entriesOS.close()
```
these really ought to be in `try ... finally` blocks to ensure we're always closing the output streams. The Oracle standard library tends to burden the containing class with resource management. Under that theory, Encoder and friends should be Closable and have their close method just call their childrens' close.

danking commented on src/main/scala/is/hail/io/RowStore.scala:
```diff
@@ -521,7 +523,95 @@ object RichRDDRegionValue {
+
+    val fullRowType = t.rvRowType
+    val rowsRVType = t.rowType
+    val localEntriesIndex: Int = t.entriesIdx
```
we tend to avoid type annotations on local vals, is there a reason for this one?

danking commented on src/main/scala/is/hail/io/RowStore.scala:
```diff
@@ -521,7 +523,95 @@ object RichRDDRegionValue {
+    entriesSpec.write(hConf, path + "/entries/rows")
+
+    val itemCount = partitionCounts.sum
+    info(s"wrote ${ partitionCounts.sum } items in $nPartitions partitions")
```
I think item count is unused

tpoterba commented on src/main/scala/is/hail/expr/Relational.scala:
```diff
@@ -227,59 +205,97 @@ case class MatrixRead(
+              while (i < fullRowType.size) {
+                rv2b.addField(fullRowType, rv, i - 1)
+                i += 1
+              }
```
it looks like this removes the `if` in favor of two while loops for performance reasons

cseed replied:
I guess it "enables a future option".  The MatrixType coming from read is parsed by the parser which constructs with `MatrixType.fromParts` but I didn't want to necessarily rely on this behavior which is rather long-range.

cseed commented:
Addressed comments and rebased.

We have a stream (and decoder) as a member of the iterator returned by `readRowsPartition`.  Right now we close when the iterator returns the last element, but ... in general I'm not sure how to clean this up properly.  It will leak if we don't iterate all the way through and I don't see a way to fix that.  I can't use finally because I want to leave it open until the end.  I can catch Exceptions coming out of next and hasNext, close, and rethrow.  And close in the finalizer.  Is that enough?  This is terrible.  Spark should really be carrying around closeable iterators.

cseed commented:
Pushed changes to clean up in readRowsPartition:

 - close in if creating encoder throws Exception,
 - close dec in finalizer,
 - close dec if body of next throws Exception

That's the best I can see.  Let me know if you have other ideas.
